### PR TITLE
Create parser strategy

### PR DESCRIPTION
**Pull request by jaimerson**
This strategy is meant to detect calls to class methods and add the referenced class to the case map.

### CODE REVIEW

File: features/modified_execution_paths/class_method_call_spec.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+describe 'Changing source file' do
+  subject(:forecast) do
+    Crystalball.foresee(workdir: root, map_path: root.join('execution_map.yml')) do |predictor|
+      predictor.use Crystalball::Predictor::ModifiedExecutionPaths.new
+    end
+  end
+  include_context 'simple git repository'
+
+  it 'adds class2 example when class1 changes because of class method call' do
+    change class1_path
```
Comment (EPecherkin) [inline features/modified_execution_paths/class_method_call_spec.rb]: This is not enough to test it. This case will be handled as well by coverage strategy

File: lib/crystalball/map_generator/parser_strategy/processor.rb [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        end
+
+        def on_casgn(node, scope)
+          namespace, name, = node.children
+          scope_name = namespace ? qualified_name(qualified_name_from_node(namespace), scope) : scope
+          add_constant_defined(name, scope_name)
+        end
+
+        def on_class(node, scope)
+          const, _, body = node.children # TODO: _ is superclass
```
Comment (EPecherkin) [inline lib/crystalball/map_generator/parser_strategy/processor.rb]: why `TODO` ? are you going to handle it in some way?

Comment (jaimerson) [reply lib/crystalball/map_generator/parser_strategy/processor.rb]: Yes, I'll add it to the map as well

Comment (EPecherkin) [reply lib/crystalball/map_generator/parser_strategy/processor.rb]: write a more detailed description in such case. Like what should be done here why it isn't

File: crystalball.gemspec [hunk 3]
```diff
@@ -31,6 +31,7 @@ Gem::Specification.new do |spec|
   spec.require_paths = ["lib"]
 
   spec.add_dependency 'git'
+  spec.add_dependency 'parser'
```
Comment (pluff) [inline crystalball.gemspec]: I think this shouldn't be a default dependency. Let's move it to development dependency only.
We need to mention in README that if you want to use parser strategy you need to add `parser` to your Gemfile.

File: features/modified_execution_paths/class_method_call_spec.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+    is_expected.to include(
+      './spec/class2_spec.rb[1:4:1]'
+    )
```
Comment (pluff) [inline features/modified_execution_paths/class_method_call_spec.rb]: Please reuse [this](https://github.com/toptal/crystalball/pull/68/files#diff-0fe6e382aea1d1671075e5319b722f6cR7) when it will be merged

File: lib/crystalball/map_generator/parser_strategy.rb [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      def files_to_inspect
+        Dir.glob(File.join(root_path, '**/*.rb')).grep(pattern)
+      end
+
+      def tracer_for(case_map)
+        TracePoint.new(:call) do |tp|
+          classes = [tp.defined_class.to_s, tp.defined_class.singleton_class.to_s]
+          used = const_definition_paths.select { |k, _| classes.include?(k) }.keys
+          case_map.push(*used)
+        end
```
Comment (pluff) [inline lib/crystalball/map_generator/parser_strategy.rb]: where this method is used?

File: lib/crystalball/map_generator/parser_strategy/processor.rb [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+          parse_and_process(path)
+          consts_defined
+        end
+
+        def consts_interacted_with_in(path)
+          self.current_scope = nil
+          self.consts_interacted_with = []
+          parse_and_process(path)
+          consts_interacted_with
+        end
```
Comment (pluff) [inline lib/crystalball/map_generator/parser_strategy/processor.rb]: why all these handlers are public?

File: lib/crystalball/map_generator/parser_strategy/processor.rb [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+require 'parser/current'
+
+module Crystalball
+  class MapGenerator
+    class ParserStrategy
+      # Parses the given source files and adds the class and module definitions
+      # to the `consts_defined` array.
+      class Processor
+        attr_accessor :consts_defined, :consts_interacted_with, :current_scope
```
Comment (pluff) [inline lib/crystalball/map_generator/parser_strategy/processor.rb]: why these are public?

**Review by pluff — CHANGES_REQUESTED** [review]
Did you test the quality of `parser` AST? Does it handle corner cases? E.g. `MyConst = Class.new` and `MyConst = Object.new`?

File: lib/crystalball/map_generator/parser_strategy.rb (hunk 5 above)
Comment (jaimerson) [reply lib/crystalball/map_generator/parser_strategy.rb]: Oh, I forgot to remove it

File: lib/crystalball/map_generator/parser_strategy/processor.rb (hunk 6 above)
Comment (jaimerson) [reply lib/crystalball/map_generator/parser_strategy/processor.rb]: You're right, they shouldn't be. They started like this because this class sublclassed `Parser::AST::Processor`, but that's no longer the case. I'll change it.

**Comment by jaimerson** [comment]
@pluff yes, it handles constant assignment as well https://github.com/toptal/crystalball/blob/ddd2f4fc79550084eb7ce7cc05a7d20267b3efd2/spec/map_generator/parser_strategy/processor_spec.rb#L110-L125
And also `CONSTANT = :foo`

File: features/modified_execution_paths/class_method_call_spec.rb (hunk 1 above)
Comment (EPecherkin) [reply features/modified_execution_paths/class_method_call_spec.rb]: This is still not enough

File: lib/crystalball.rb [hunk 8]
```diff
@@ -14,6 +14,7 @@
 require 'crystalball/map_generator/configuration'
 require 'crystalball/map_generator/coverage_strategy'
 require 'crystalball/map_generator/allocated_objects_strategy'
+require 'crystalball/map_generator/parser_strategy'
```
Comment (EPecherkin) [inline lib/crystalball.rb]: I think parser strategy should be included separately in favor of https://github.com/toptal/crystalball/pull/66/files#r185799760

File: lib/crystalball/map_generator/parser_strategy/processor.rb (hunk 2 above)
Comment (EPecherkin) [reply lib/crystalball/map_generator/parser_strategy/processor.rb]: This one still needs a description

File: features/modified_execution_paths/class_method_call_spec.rb (hunk 1 above)
Comment (jaimerson) [reply features/modified_execution_paths/class_method_call_spec.rb]: @EPecherkin I changed `Class1.foo` to be an empty method and confirmed it is not caught by coverage or any of the other strategies.

**Review by pluff — APPROVED** [review]

**Review by EPecherkin — APPROVED** [review]
